### PR TITLE
[jvm-packages][breaking]Rework xgboost4j jvm packages for spark

### PR DESCRIPTION
**Pull request by wbo4958**
# About this PR

## Features and Fixes

- Introduce an abstract XGBoost Estimator:
  - Create an abstract base class that encapsulates common functionality for all XGBoost estimators.
  - This will promote code reuse and maintain consistency across different estimators.

- Update to the latest XGBoost parameters
  - Add all XGBoost parameters supported in XGBoost4j-spark base on [this](https://xgboost.readthedocs.io/en/stable/parameter.html)
  - Add setter and getter for these parameters.
  - Remove the deprecated parameters

- Implement support for XGBoostRanker:
  - Add a new XGBoostRanker estimator for ranking problem.

- Address the missing value handling:
  - 0 is a meaningful value in machine learning case, the old version enforces 0 as the missing value which may result in XGBoost model inaccurate.

- Consolidate xgboost4j-gpu and xgboost4j-spark-gpu:
  - Move the functionality of xgboost4j-gpu into xgboost4j-spark-gpu.
  - Remove any soft links in xgboost4j-spark-gpu to ensure a clean structure, Instead, make xgboost4j-spark-gpu extend from xgboost4j-spark and xgboost4j

- Remove any ETL operations in XGBoost
  - As a thin wrapper of XGBoost4j, it shouldn't wrap much ETL in XGBoost4j itself to make code hard to maintain and read.

- Create uber jars for xgboost4j-spark and xgboost4j-spark-gpu:
  - Create aggregator module to build uber jars (fat jars) that include all the necessary dependencies for xgboost4j-spark
  - Create aggregator-gpu module to build uber jars for xgboost4j-spark-gpu.

- Rework the GPU plugin:
  - Rework the GPU plugin to improve its stability and make it more readable and maintainable.

- Expand sanity tests for CPU and GPU consistency:
  - Increase the coverage of sanity tests to verify that the training and transform results are consistent between CPU and GPU implementations.
  - This will help identify and prevent any discrepancies or bugs related to the GPU support.

## Breaking changes

- Remove some unused parameters like Rabit related parameters and train_test_ratio. etc.
- Separate XGBoost-spark parameter from the whole XGBoost parameters, and make the constructor of XGBoost estimator only accept the XGBoost parameters instead of parameters for xgboost4j-spark like,
  ``` scala
  val xgbParams: Map[String, Any] = Map(
        "max_depth" -> 5,
        "eta" -> 0.2,
        "objective" -> "binary:logistic"
      )
  val est = new XGBoostClassifier(xgbParams)
        .setNumRound(10)
        .setDevice("cpu")
        .setNumWorkers(2)
  ``` 
   or

    ``` scala
    val est = new XGBoostClassifier(xgbParams)
          .setMaxDepth(5)
          .setEta(0.2)
          .setObjective("binary:logistic")
          .setNumRound(10)
          .setDevice("cpu")
          .setNumWorkers(2)
    ```
- XGBoostRegressor doesn't support ranking problem anymore, the functionality of ranking has been moved to the new estimator XGBoostRanker. And remove the ETL (like grouping) for ranking problem.
- Removed the xgboost4j-gpu module and move the functionality into xgboost4j-spark-gpu
 

## Test this PR

- Unit tests For scala 2.12

``` bash
mvn clean package
mvn -P gpu clean package
```

- Unit tests For scala 2.13
``` bash
cd xgboost
python dev/change_scala_version.py --scala-version 2.13

cd jvm-packages
mvn clean package
mvn -P gpu clean package
```

### CODE REVIEW

**Comment by wbo4958** [comment]
hey @trivialfis @hcho3 @dotbg, please help review this PR, thx very much.

**Comment by hcho3** [comment]
Would you please split this pull requests into smaller chunks so that it is easier to review?

**Comment by wbo4958** [comment]
It's hard to separate it, since it's completely jvm rewrite. The whole code architecture has been re-orged. and totally different implementation.

**Comment by hcho3** [comment]
I won't review the code, since I don't think I am qualified to review such a large change in the Java/Scala code base. The best I can do is to review the changes in the CI/CD pipelines.

**Comment by wbo4958** [comment]
> I won't review the code, since I don't think I am qualified to review such a large change in the Java/Scala code base. The best I can do is to review the changes in the CI/CD pipelines.

Thx

**Comment by dotbg** [comment]
@wbo4958  thank you for the PR. I shall find some time to go through it. Unfortunately, I was very busy lately :( 
@hcho3  thank you for reviewing the CI/CD part. this should easen the review for me.

File: jvm-packages/pom.xml [hunk 1]
```diff
@@ -2,511 +2,528 @@
...
+    <connection>scm:git:git:/github.com/dmlc/xgboost.git</connection>
+    <developerConnection>scm:git:ssh://github.com/dmlc/xgboost.git</developerConnection>
+    <url>https://github.com/dmlc/xgboost</url>
+  </scm>
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+    <maven.compiler.source>1.8</maven.compiler.source>
+    <maven.compiler.target>1.8</maven.compiler.target>
+    <flink.version>1.19.0</flink.version>
```
Comment (dotbg) [inline jvm-packages/pom.xml:35]:
```suggestion
    <flink.version>1.19.1</flink.version>
```

File: jvm-packages/pom.xml [hunk 2]
```diff
@@ -2,511 +2,528 @@
...
+    <junit.version>4.13.2</junit.version>
+    <spark.version>3.5.1</spark.version>
+    <spark.version.gpu>3.5.1</spark.version.gpu>
+    <fasterxml.jackson.version>2.15.2</fasterxml.jackson.version>
```
Comment (dotbg) [inline jvm-packages/pom.xml:39]:
```suggestion
    <fasterxml.jackson.version>2.17.2</fasterxml.jackson.version>
```

File: jvm-packages/pom.xml [hunk 3]
```diff
@@ -2,511 +2,528 @@
...
+    <scala.version>2.12.18</scala.version>
+    <scala.binary.version>2.12</scala.binary.version>
+    <hadoop.version>3.4.0</hadoop.version>
+    <maven.wagon.http.retryHandler.count>5</maven.wagon.http.retryHandler.count>
+    <log.capi.invocation>OFF</log.capi.invocation>
+    <use.cuda>OFF</use.cuda>
+    <spark.rapids.version>24.04.1</spark.rapids.version>
```
Comment (dotbg) [inline jvm-packages/pom.xml:46]:
```suggestion
    <spark.rapids.version>24.06.0</spark.rapids.version>
```

File: jvm-packages/pom.xml [hunk 4]
```diff
@@ -2,511 +2,528 @@
...
+    <cudf.classifier>cuda12</cudf.classifier>
+    <scalatest.version>3.2.18</scalatest.version>
```
Comment (dotbg) [inline jvm-packages/pom.xml:48]:
```suggestion
    <scalatest.version>3.2.19</scalatest.version>
```

File: jvm-packages/pom.xml [hunk 5]
```diff
@@ -2,511 +2,528 @@
...
-                </executions>
-            </plugin>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-surefire-plugin</artifactId>
-                <version>3.3.1</version>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-jar-plugin</artifactId>
+            <version>3.4.1</version>
```
Comment (dotbg) [inline jvm-packages/pom.xml:121]:
```suggestion
            <version>3.4.2</version>
```

File: jvm-packages/scalastyle-config.xml [hunk 6]
```diff
@@ -82,19 +82,27 @@ This file is divided into 3 sections:
   </check>
 
   <check level="error" class="org.scalastyle.scalariform.ClassNamesChecker" enabled="true">
-    <parameters><parameter name="regex"><![CDATA[[A-Z][A-Za-z]*]]></parameter></parameters>
```
Comment (dotbg) [inline jvm-packages/scalastyle-config.xml:85]: let's have these changes in another PR.

File: jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkMLlibPipeline.scala [hunk 7]
```diff
@@ -87,11 +88,9 @@ object SparkMLlibPipeline {
         "max_depth" -> 2,
         "objective" -> "multi:softprob",
         "num_class" -> 3,
-        "num_round" -> 100,
-        "num_workers" -> numWorkers,
         "device" -> device
       )
-    )
+    ).setNumRound(10).setNumWorkers(numWorkers)
```
Comment (dotbg) [inline jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkMLlibPipeline.scala:93]: will it still work if we specify the map values instead?

**Review by dotbg — COMMENTED** [review]
left a few comments, but I am rather far from done yet

**Comment by wbo4958** [comment]
HI @dotbg, Thx for review, some other reviewers complained this PR is quite huge PR, So I will break it down to smaller PRs. Will let you know. Thx again.

**Comment by trivialfis** [comment]
@wbo4958 can we close this PR?

**Comment by wbo4958** [comment]
Yes, close this PR

**Pull request closed without merging**